src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp

```diff
@@ -196,19 +205,229 @@ Scalar<DataType> time_deriv_of_lapse(
+  }
+  // get \partial_0 N^j
+  auto inverse_spatial_metric = determinant_and_inverse(spatial_metric).second;
+  auto dt_shift = time_deriv_of_shift<SpatialDim, Frame, DataType>(
+      lapse, shift, inverse_spatial_metric, spacetime_unit_normal, phi, pi);
+  // Following code (till the end of next loop) to be replaced with:
+  // auto dt_spatial_metric =
+  //  GeneralizedHarmonic::time_deriv_of_spatial_metric<SpatialDim, Frame,
+  //    DataType>(lapse, shift, phi, pi);
+  tnsr::ii<DataType, SpatialDim, Frame> dt_spatial_metric(get(lapse));
+  for (size_t i = 0; i < SpatialDim; ++i) {
+    for (size_t j = i; j < SpatialDim; ++j) {
+      dt_spatial_metric.get(i, j) = -get(lapse) * pi.get(i + 1, j + 1);
+      for (size_t k = 0; k < SpatialDim; ++k) {
+        dt_spatial_metric.get(i, j) += shift.get(k) * phi.get(k, i + 1, j + 1);
+      }
+    }
+  }
+  for (size_t i = 0; i < SpatialDim; ++i) {
+    dt_lower_shift->get(i) = spatial_metric.get(i, 0) * dt_shift.get(0) +
+                             shift.get(0) * dt_spatial_metric.get(i, 0);
+    for (size_t j = 0; j < SpatialDim; ++j) {
+      if (j != 0) {
+        dt_lower_shift->get(i) += spatial_metric.get(i, j) * dt_shift.get(j) +
+                                  shift.get(j) * dt_spatial_metric.get(i, j);
+      }
+    }
+  }
+}
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+tnsr::i<DataType, SpatialDim, Frame> time_deriv_of_lower_shift(
+    const Scalar<DataType>& lapse,
+    const tnsr::I<DataType, SpatialDim, Frame>& shift,
+    const tnsr::ii<DataType, SpatialDim, Frame>& spatial_metric,
+    const tnsr::A<DataType, SpatialDim, Frame>& spacetime_unit_normal,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi,
+    const tnsr::aa<DataType, SpatialDim, Frame>& pi) noexcept {
+  tnsr::i<DataType, SpatialDim, Frame> dt_lower_shift{};
+  GeneralizedHarmonic::time_deriv_of_lower_shift<SpatialDim, Frame, DataType>(
+      make_not_null(&dt_lower_shift), lapse, shift, spatial_metric,
+      spacetime_unit_normal, phi, pi);
+  return dt_lower_shift;
+}
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+void spacetime_deriv_of_norm_of_shift(
+    const gsl::not_null<tnsr::a<DataType, SpatialDim, Frame>*> d4_norm_of_shift,
+    const Scalar<DataType>& lapse,
+    const tnsr::I<DataType, SpatialDim, Frame>& shift,
+    const tnsr::ii<DataType, SpatialDim, Frame>& spatial_metric,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
+    const tnsr::AA<DataType, SpatialDim, Frame>& inverse_spacetime_metric,
+    const tnsr::A<DataType, SpatialDim, Frame>& spacetime_unit_normal,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi,
+    const tnsr::aa<DataType, SpatialDim, Frame>& pi) noexcept {
+  if (UNLIKELY(get_size(get<0>(*d4_norm_of_shift)) != get_size(get(lapse)))) {
+    *d4_norm_of_shift = tnsr::a<DataType, SpatialDim, Frame>(get(lapse));
+  }
+  auto lower_shift = raise_or_lower_index(shift, spatial_metric);
```

Comment:
These should all be done using a Variables and not-null approach to avoid the many allocations.